File: repository/src/main/java/by/rekuts/travelagency/repository/CrudRepository.java
package by.rekuts.travelagency.repository;

import java.io.Serializable;
import java.util.List;

public interface CrudRepository<T extends Serializable> {
    /**
     * This method inserts object of T class to database
     * @param t - object
     */
    void insert(T t);

    /**
     * This method updates object of T class in database
     * @param t - object
     */
    void update(T t);

    /**
     * This method deletes object with id <b>id</b> from database
     * @param id - id of the object
     */
    void delete(int id);

    /**
     * @return list of all objects from database or by id
     */
    List<T> getList(Specification specification);
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/CrudRepository.scala
package by.rekuts.travelagency.repository

import java.util

import javax.persistence.{EntityManager, PersistenceContext}

trait CrudRepository[T <: java.io.Serializable] {
  @PersistenceContext
  var entityManager: EntityManager = _
  /**
    * This method inserts object of T class to database
    *
    * @param t - object
    */
  def insert(t: T): Unit = entityManager.persist(t)

  /**
    * This method updates object of T class in database
    *
    * @param t - object
    */
  def update(t: T): Unit

  /**
    * This method deletes object with id <b>id</b> from database
    *
    * @param id - id of the object
    */
  def delete(id: Int): Unit

  /**
    * @return list of all objects from database or by id
    */
  def getList(specification: Specification[_]): util.List[T]
}
